### ISSUE TITLE
PodMonitor matchExpression selector broken after 0.79

### ISSUE THREAD

**Issue by tommyzli** [prometheus-operator/prometheus-operator#7214]
### Is there an existing issue for this?

- [X] I have searched the existing issues

## Description
Version 0.79.0 breaks podmonitors using the `DoesNotExist` matchExpression.

## Steps to Reproduce
Create a pod monitor:
```
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  name: test-podmonitor
spec:
  jobLabel: test-stats
  namespaceSelector:
    any: true
  podMetricsEndpoints:
  - interval: 30s
    path: /metrics
  selector:
    matchExpressions:
    - key: any-key
      operator: DoesNotExist
```
Then run `kubectl describe podmonitor test-podmonitor`
## Expected Result
nothing, it should be a valid podmonitor

## Actual Result
The events show:
```
  Warning  InvalidConfiguration  15s   prometheus-controller  PodMonitor test-podmonitor was rejected due to invalid configuration: failed to create label requirement: operator: Unsupported value: "doesnotexist": supported values: "in", "notin", "=", "==", "!=", "gt", "lt", "exists", "!"
```

### Prometheus Operator Version

```shell
quay.io/prometheus-operator/prometheus-operator:v0.79.1
```

### Kubernetes Version

```yaml
clientVersion:
  buildDate: "2024-03-14T23:58:34Z"
  compiler: gc
  gitCommit: 6813625b7cd706db5bc7388921be03071e1a492d
  gitTreeState: clean
  gitVersion: v1.29.3
  goVersion: go1.21.8
  major: "1"
  minor: "29"
  platform: darwin/arm64
kustomizeVersion: v5.0.4-0.20230601165947-6ce0bf390ce3
serverVersion:
  buildDate: "2024-10-23T23:40:34Z"
  compiler: gc
  gitCommit: d06fe5fa1c524ee5eb5d16bb9edc4dda6075f6c3
  gitTreeState: clean
  gitVersion: v1.29.10-eks-7f9249a
  goVersion: go1.22.8
  major: "1"
  minor: 29+
  platform: linux/amd64
```

### Kubernetes Cluster Type

EKS

### How did you deploy Prometheus-Operator?

helm chart:prometheus-community/kube-prometheus-stack

### Manifests

```yaml
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  name: envoy-stats-monitor
  namespace: istio-system
  labels:
    monitoring: istio-proxies
    release: istio
spec:
  selector:
    matchExpressions:
    - {key: istio-prometheus-ignore, operator: DoesNotExist}
  namespaceSelector:
    any: true
  jobLabel: envoy-stats
  podMetricsEndpoints:
  - path: /stats/prometheus
    interval: 15s
    relabelings:
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_container_name]
      regex: "istio-proxy"
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_annotationpresent_prometheus_io_scrape]
    - sourceLabels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
      action: replace
      regex: ([^:]+)(?::\d+)?;(\d+)
      replacement: $1:$2
      targetLabel: __address__
    - action: labeldrop
      regex: "__meta_kubernetes_pod_label_(.+)"
    - sourceLabels: [__meta_kubernetes_namespace]
      action: replace
      targetLabel: namespace
    - sourceLabels: [__meta_kubernetes_pod_name]
      action: replace
      targetLabel: pod_name
```

### prometheus-operator log output

```shell
ts=2024-12-17T21:08:11.783532321Z level=warn caller=/workspace/pkg/prometheus/resource_selector.go:460 msg="skipping podmonitor" component=prometheus-controller error="failed to create label requirement: operator: Unsupported value: \"doesnotexist\": supported values: \"in\", \"notin\", \"=\", \"==\", \"!=\", \"gt\", \"lt\", \"exists\", \"!\"" podmonitor=istio-system/envoy-stats-monitor namespace=lighthouse prometheus=kube-prometheus-stack-prometheus
ts=2024-12-17T21:08:11.82246845Z level=info caller=/go/pkg/mod/k8s.io/client-go@v0.32.0/tools/record/event.go:389 msg="Event occurred" object.name=envoy-stats-monitor object.namespace=istio-system fieldPath="" kind=PodMonitor apiVersion=monitoring.coreos.com/v1 type=Warning reason=InvalidConfiguration message="PodMonitor envoy-stats-monitor was rejected due to invalid configuration: failed to create label requirement: operator: Unsupported value: \"doesnotexist\": supported values: \"in\", \"notin\", \"=\", \"==\", \"!=\", \"gt\", \"lt\", \"exists\", \"!\""
```

### Anything else?

The full monitor spec is defined here https://github.com/istio/istio/blob/0c4570a10cb012a6f63796dcc6ec72d83599f1c7/samples/addons/extras/prometheus-operator.yaml

**Comment by tommyzli** [comment]
I pushed a branch here with a unit test change that reveals the panic in the validation: https://github.com/tommyzli/prometheus-operator/tree/broken-matchexpression-test
```
$ go test -run ^TestPodMonitorSelectors$ github.com/prometheus-operator/prometheus-operator/pkg/prometheus
--- FAIL: TestPodMonitorSelectors (0.00s)
    --- FAIL: TestPodMonitorSelectors/PodMonitor_with_Match_Expression_Selector (0.00s)
panic: failed to create label requirement: operator: Unsupported value: "doesnotexist": supported values: "in", "notin", "=", "==", "!=", "gt", "lt", "exists", "!" [recovered]
	panic: failed to create label requirement: operator: Unsupported value: "doesnotexist": supported values: "in", "notin", "=", "==", "!=", "gt", "lt", "exists", "!"

goroutine 30 [running]:
testing.tRunner.func1.2({0x1047cd220, 0x1400054fa20})
	/Users/tommyli/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.3.darwin-arm64/src/testing/testing.go:1632 +0x1bc
testing.tRunner.func1()
	/Users/tommyli/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.3.darwin-arm64/src/testing/testing.go:1635 +0x334
panic({0x1047cd220?, 0x1400054fa20?})
	/Users/tommyli/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.3.darwin-arm64/src/runtime/panic.go:785 +0x124
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ConfigGenerator).generateRoleSelectorConfig(0x14000174900, {0x140004bfb00, 0x2, 0x2}, {0x14000050f98, 0x1, 0x0?}, {0x0?, {0x140004bf600?, 0x10429afdc?, ...}})
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg.go:2237 +0x608
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ConfigGenerator).generatePodMonitorConfig.(*ConfigGenerator).withK8SRoleSelectorConfig.func1({0x140004bfb00, 0x2, 0x2})
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg.go:2135 +0x114
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ConfigGenerator).generateK8SSDConfig(0x14000174900, {0xc0?, {0x0?, 0x2?, 0x0?}}, {0x104324bc2, 0x7}, 0x0, {0x104a3a3a0, 0x1400054f9c0}, ...)
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg.go:2219 +0x87c
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ConfigGenerator).generatePodMonitorConfig(_, _, {0x140004ed3f0, 0x0, 0x0, {0x0, 0x0}, {0x0, 0x0}, 0x0, ...}, ...)
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg.go:1231 +0x4b0
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ConfigGenerator).appendPodMonitorConfigs(0x140001747e0, {0x0, 0x0, 0x0}, 0x14000051f28, 0x0, 0x140004bfa00, 0x1)
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg.go:2882 +0x200
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ConfigGenerator).GenerateServerConfiguration(0x140001747e0, 0x1400040aa08, 0x0, 0x14000051f28, 0x0, 0x0, 0x140004bfa00, {0x0, 0x0, 0x0}, ...)
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg.go:863 +0x314
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.TestPodMonitorSelectors.func1(0x1400062cb60)
	/Users/tommyli/src/prometheus-operator/pkg/prometheus/promcfg_test.go:12630 +0x2f8
testing.tRunner(0x1400062cb60, 0x14000700690)
	/Users/tommyli/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.3.darwin-arm64/src/testing/testing.go:1690 +0xe4
created by testing.(*T).Run in goroutine 28
	/Users/tommyli/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.3.darwin-arm64/src/testing/testing.go:1743 +0x314
FAIL	github.com/prometheus-operator/prometheus-operator/pkg/prometheus	0.545s
FAIL
```
Looks like the failing codepath was introduced here https://github.com/prometheus-operator/prometheus-operator/pull/7185

**Comment by nicolastakashi** [comment]
Hmmm!
Interesting, looks like some inconsistency on K8S Labels package not being able to parse doesnotexist to one of the following operators

**Comment by nicolastakashi** [comment]
I've been looking into the code, and it looks like the https://github.com/kubernetes/apimachinery/blob/v0.32.0/pkg/selection/operator.go#L21 doesn't know anything about the `doesnotexist` this operator on that package is represented by `!`


I'll open a PR addressing the issue

**Comment by slashpai** [comment]
IIUC we don't need to convert to lowercase in https://github.com/prometheus-operator/prometheus-operator/blob/25e59cb0b8b0563cabfbd7e0f1464d0e0dd10672/pkg/prometheus/resource_selector.go#L390?

**Comment by slashpai** [comment]
It looks to be due to the change https://github.com/kubernetes/apimachinery/commit/cc2ba35af57ca67ba4acdbbe17c10a5faaaafdd5, we would need to update validateMatchExpression()

**Comment by simonpasquier** [comment]
Good catch! We'd need a switch/case here to translate from the user-visible value to the internal enum value https://github.com/prometheus-operator/prometheus-operator/blob/25e59cb0b8b0563cabfbd7e0f1464d0e0dd10672/pkg/prometheus/promcfg.go#L2234-L2240

**Comment by nicolastakashi** [comment]
PR addressing the issue: https://github.com/prometheus-operator/prometheus-operator/pull/7216

**Comment by slashpai** [comment]
Fix available in v0.79.2 https://github.com/prometheus-operator/prometheus-operator/releases/tag/v0.79.2
